File: src/test/java/io/veridia/jsonlogic/StringExpressionTests.java
package io.veridia.jsonlogic;

import com.fasterxml.jackson.core.JsonProcessingException;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class StringExpressionTests {

    private static final JsonLogic jsonLogic = new JsonLogic();

    // -------------------------------------------------
    // contains
    // -------------------------------------------------

    @Test
    public void testContainsTrue() throws JsonProcessingException {
        String json = "{\"contains\": [\"hello world\", \"world\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(true, result);
    }

    @Test
    public void testContainsFalse() throws JsonProcessingException {
        String json = "{\"contains\": [\"hello\", \"xyz\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(false, result);
    }

    // -------------------------------------------------
    // starts_with
    // -------------------------------------------------

    @Test
    public void testStartsWithTrue() throws JsonProcessingException {
        String json = "{\"starts_with\": [\"abcdef\", \"abc\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(true, result);
    }

    @Test
    public void testStartsWithFalse() throws JsonProcessingException {
        String json = "{\"starts_with\": [\"abcdef\", \"bcd\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(false, result);
    }

    // -------------------------------------------------
    // ends_with
    // -------------------------------------------------

    @Test
    public void testEndsWithTrue() throws JsonProcessingException {
        String json = "{\"ends_with\": [\"abcdef\", \"def\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(true, result);
    }

    @Test
    public void testEndsWithFalse() throws JsonProcessingException {
        String json = "{\"ends_with\": [\"abcdef\", \"abc\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(false, result);
    }

    // -------------------------------------------------
    // regex_match (RE2)
    // -------------------------------------------------

    @Test
    public void testRegexMatchSimple() throws JsonProcessingException {
        String json = "{\"regex_match\": [\"google.com\", \"google\\\\.com\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(true, result);
    }

    @Test
    public void testRegexMatchFindNotFullMatch() throws JsonProcessingException {
        String json = "{\"regex_match\": [\"https://google.com/search\", \"google\\\\.com\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(true, result);
    }

    @Test
    public void testRegexMatchNoMatch() throws JsonProcessingException {
        String json = "{\"regex_match\": [\"example.com\", \"google\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(false, result);
    }

    @Test
    public void testRegexInvalidPattern() throws JsonProcessingException {
        String json = "{\"regex_match\": [\"test\", \"(\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(false, result);
    }

    // -------------------------------------------------
    // safety limits
    // -------------------------------------------------

    @Test
    public void testRegexPatternTooLong() throws JsonProcessingException {
        String longPattern = "a".repeat(300);
        String json = "{\"regex_match\": [\"aaa\", \"" + longPattern + "\"]}";

        Object result = jsonLogic.apply(json, null);
        assertEquals(false, result);
    }

    @Test
    public void testRegexInputTooLong() throws JsonProcessingException {
        String longInput = "a".repeat(2000);
        String json = "{\"regex_match\": [\"" + longInput + "\", \"a+\"]}";

        Object result = jsonLogic.apply(json, null);
        assertEquals(false, result);
    }

    // -------------------------------------------------
    // type coercion & null handling
    // -------------------------------------------------

    @Test
    public void testContainsWithNumber() throws JsonProcessingException {
        String json = "{\"contains\": [12345, \"234\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(true, result);
    }

    @Test
    public void testContainsNullLeft() throws JsonProcessingException {
        String json = "{\"contains\": [null, \"a\"]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(false, result);
    }

    @Test
    public void testContainsNullRight() throws JsonProcessingException {
        String json = "{\"contains\": [\"abc\", null]}";
        Object result = jsonLogic.apply(json, null);

        assertEquals(false, result);
    }
}

File: src/test/scala/io/veridia/jsonlogic/StringExpressionTests.scala
package io.veridia.jsonlogic

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

import java.util as ju

class StringExpressionTests extends AnyFunSuite with Matchers:
  private val jsonLogic = new JsonLogic()

  private val cases: Seq[(String, String, Boolean)] = Seq(
    ("contains: true", "{\"contains\": [\"hello world\", \"world\"]}", true),
    ("contains: false", "{\"contains\": [\"hello\", \"xyz\"]}", false),
    ("starts_with: true", "{\"starts_with\": [\"abcdef\", \"abc\"]}", true),
    ("starts_with: false", "{\"starts_with\": [\"abcdef\", \"bcd\"]}", false),
    ("ends_with: true", "{\"ends_with\": [\"abcdef\", \"def\"]}", true),
    ("ends_with: false", "{\"ends_with\": [\"abcdef\", \"abc\"]}", false),
    ("regex_match: simple", "{\"regex_match\": [\"google.com\", \"google\\\\.com\"]}", true),
    ("regex_match: find, not full match", "{\"regex_match\": [\"https://google.com/search\", \"google\\\\.com\"]}", true),
    ("regex_match: no match", "{\"regex_match\": [\"example.com\", \"google\"]}", false),
    ("regex_match: invalid pattern fails closed", "{\"regex_match\": [\"test\", \"(\"]}", false),
    ("contains: number coerces to its string form", "{\"contains\": [12345, \"234\"]}", true),
    ("contains: null left is false", "{\"contains\": [null, \"a\"]}", false),
    ("contains: null right is false", "{\"contains\": [\"abc\", null]}", false),
    ("contains: empty substring is true", "{\"contains\": [\"abc\", \"\"]}", true),
    ("starts_with: empty prefix is true", "{\"starts_with\": [\"abc\", \"\"]}", true),
    ("contains: on an empty string is false", "{\"contains\": [\"\", \"a\"]}", false),
    ("regex_match: empty input matches .*", "{\"regex_match\": [\"\", \".*\"]}", true),
    ("regex_match: anchors", "{\"regex_match\": [\"/pricing\", \"^/pricing$\"]}", true),
    ("contains: unicode", "{\"contains\": [\"привет мир\", \"мир\"]}", true)
  )

  for (name, json, expected) <- cases do
    test(name) {
      jsonLogic.apply(json, null) shouldEqual expected
    }

  test("regex_match: pattern over the 256-char safety limit fails closed") {
    val longPattern = "a".repeat(300)
    jsonLogic.apply(s"""{"regex_match": ["aaa", "$longPattern"]}""", null) shouldEqual false
  }

  test("regex_match: input over the 1024-char safety limit fails closed") {
    val longInput = "a".repeat(2000)
    jsonLogic.apply(s"""{"regex_match": ["$longInput", "a+"]}""", null) shouldEqual false
  }

  test("contains: resolves left operand via a context variable") {
    val ctx = ju.Map.of("request", ju.Map.of("path", "/pricing/plans"))
    jsonLogic.apply("{ \"contains\": [ {\"var\": \"request.path\"}, \"/pricing\" ] }", ctx) shouldEqual true
  }

  test("regex_match: resolves left operand via a context variable") {
    val ctx = ju.Map.of("referrer", ju.Map.of("host", "www.google.com"))
    jsonLogic.apply("{ \"regex_match\": [ {\"var\": \"referrer.host\"}, \"google\\\\.com$\" ] }", ctx) shouldEqual true
  }

  private object NullToString:
    override def toString: String = null

  test("contains: a non-null value whose own toString() returns null is false, not an NPE") {
    val ctx = ju.Collections.singletonMap("x", NullToString: Object)
    jsonLogic.apply("{\"contains\": [{\"var\": \"x\"}, \"abc\"]}", ctx) shouldEqual false
  }

  test("regex_match: a non-null value whose own toString() returns null is false, not an NPE") {
    val ctx = ju.Collections.singletonMap("x", NullToString: Object)
    jsonLogic.apply("{\"regex_match\": [{\"var\": \"x\"}, \"a+\"]}", ctx) shouldEqual false
  }
